### ISSUE TITLE
LM1B example does not handle partial batches in evals

### ISSUE THREAD

**Issue by georgedahl** [google/flax#264]
Provide as much information as possible. At least, this should include a description of your issue and steps to reproduce the problem. If possible also provide a summary of what steps or workarounds you have already tried.

### Problem you have encountered:

The LM1B example calls input_poipeline.get_lm1b_datasets https://github.com/google/flax/blob/master/examples/lm1b/train.py#L350 with drop_remainder at its default of True, dropping partial batches in the validation data caused by the number of examples being evenly divisible by the batch size. This will ignore some validation data and if a user modifies this example to evaluate on their own test set it will produce incorrect results.

### What you expected to happen:

The example should be able to correctly handle any number of validation examples and drop none of them by default. This might require padding and masking. It is fine to have slightly different behavior for the training set, but for evals every example in the eval set should be processed exactly once and the metrics should be aggregated across all the eval batches correctly.

**Comment by jheek** [comment]
It's quite common to drop the remainder in for example the official TF examples. But I agree we should really do better. For a single host the fix is trivial but for multiple hosts we need to be careful about feeding the same number of batches and padding up to the same size on each host.

**Comment by georgedahl** [comment]
It isn't safe to do work intended for scientific publication with code that doesn't compute correct evals. It is a shame incorrect evals have become normalized by bad examples in other frameworks.

I know you aren't really disagreeing with me here, but I worry your comment could be read as minimizing this issue. This is a severe correctness problem even though it might be a common problem. The longer it exists in an example the more risk there is someone will put wrong eval numbers in their publications and it will spread to more codebases.

I also think this is a great example of where our tools push us to write bad code because writing correct code is inconvenient and uglier. It should be easier to handle these things which is why I additionally filed https://github.com/google/jax/issues/3088
